### PR TITLE
Pbh

### PR DESCRIPTION
**Pull request by Kerianne28**
This branch implements everything needed to inject primordial black holes (PBHs) into the .h5 file that will be used by `calc_events`.

### **Synthetic.py Changes**
_**`add_pbh`:**_
This function takes the .h5 file that is output by `perform_pop_syn`, creates everything needed for the injected PBHs, then combines the PBH data with the data from the previous .h5 file into a new .h5 file.

This function does the following:

- Reads in the .h5 file without PBHs. Maximum and minimum l and b values are determined for masking the data later.

- Uses an NFW profile to determine how many PBHs there are within 8.3kpc of the galactic center.

- Populates entire sphere with PBHs depending on the masses and fraction dark matter given.

- Determines radii values for each PBH.

- Converts radii from spherical galactocentric to galactic coordinates (and assigns longitude and latitude values).

- Masks the data, giving you only the PBHs in the field of view you need.

- Calculates the Maxwellian velocity distribution for each radius. 

- Randomly samples one value, and assigns it as the rms velocity (then converted to vx, vy, and vz).

- All other values needed are assigned to each PBH: magnitudes, pop_id, etc.

- A new .h5 file is created that consists of all of the data from the original .h5 file (without PBHs. From `perform_pop_syn`), combined with all of the new PBH information for the current field.

_**`generate_pbh_config_file`:**_

- Generates a config file with the parameters that are needed for the `add_pbh` function.

**Note:** I am not actually sure if this config file is needed. I followed the layout of other PopSyCLE functions in which some of the parameters that are set when you call the function, are the same as the parameters in the config.yaml file. It may be easier to only have a config file and not put the parameters in the call to the function, or vice versa. For now, I followed the style and created the .yaml file and put the parameters in the call to the function as well.

### **Run.py changes**

- Created an optional `add_argument` statement for the PBH config file.

- Added a segment that instructs what to do if a pbh_config file is given.

### **Other files added**
**_pbh_config.yaml:_**

- Has all of the default parameters needed for the `add_pbh` function.

**_radial_velocity_dispersion_digitized_.csv:_**

- Stores data points that are needed in the PBH velocity determination stage. It is a list of radii and dispersion points that are interpolated for all of the PBH radii, to assign each their own dispersion value. This dispersion value then is used to get f(v), which is used to generate the Maxwellian distribution in which the PBH velocity is drawn from.

### CODE REVIEW

File: popsycle/run.py [hunk 1]
```diff
@@ -103,7 +103,14 @@ def run():
...
                           help="Overwrite all output files.",
                           action='store_true')
-    args = parser.parse_args()
+
+
+    optional.add_argument('--pbh-config-filename', type=str,
+                      help='Name of configuration file containing '
+                           'pbh inputs. Default if needed is: '
+                           'pbh_config.yaml',
+                      default=None)
```
Comment (MichaelMedford) [inline popsycle/run.py]: This needs to have a default value of 'pbh_config.yaml', as described in the help string.

File: popsycle/synthetic.py [hunk 2]
```diff
@@ -3512,3 +3517,389 @@ def generate_slurm_scripts(slurm_config_filename, popsycle_config_filename,
...
+    #Defining galactocentric coordinates
+    c = coord.Galactocentric(x=cart[0] * units.kpc, y=cart[1] * units.kpc, z=cart[2] * units.kpc)
+
+    #Transforming from galactocentric to galactic coordinates
+    #outputs l, b, dist in degrees.
+    c = c.transform_to(coord.Galactic(representation_type='cartesian'))
+
+    #Defining function to use in PBH mask (this portion of the mask creates circular shape 
+    #from a rectangular shape because our field of view is circular)
+    def angdist(ra1, dec1, ra2, dec2):
```
Comment (MichaelMedford) [inline popsycle/synthetic.py]: Should this perhaps be outside of the function?

File: popsycle/synthetic.py [hunk 3]
```diff
@@ -3512,3 +3517,389 @@ def generate_slurm_scripts(slurm_config_filename, popsycle_config_filename,
...
+
+    return
+
+def generate_pbh_config_file(config_filename, galaxia_area, fdm, pbh_mass, r_max, c, r_vir):
+    """
+    Save PBH configuration parameters into a yaml file
+
+    Parameters
+    ----------
+    galaxia_area : float
```
Comment (MichaelMedford) [inline popsycle/synthetic.py]: This should be read in from the popsycle_config file that is already provided

**Comment by MichaelMedford** [comment]
Please commit an example PBH_config file into the popsycle/data directory

**Comment by MichaelMedford** [comment]
Additional comments that we just addressed via video chat:

- Master has to be pulled into the pbh branch before proceeding
- `seed` has to be passed to `add_pbh` along with the `np.random.seed(seed)` command
- `pbh_config_filename` needs to be added to arguments of `generate_slurm_scripts`, and included in the optional parameters section to be passed to the `slurm_template`

File: popsycle/synthetic.py [hunk 4]
```diff
@@ -1727,7 +1732,7 @@ def _calc_blends(bigpatch, c, event_lbt, blend_rad):
 
     # Query ball against the existing (cached) tree.
     # NOTE: results is an array of lists.
-    results = kdtree_cache.query_ball_point(flatxyz1.T, r_kdt)
+    results = kdtree_cache.query_ball_point(flatxyz1.T.copy(order='C'), r_kdt)
```
Comment (MichaelMedford) [inline popsycle/synthetic.py]: Why was this change necessary? What does it do?

File: popsycle/synthetic.py [hunk 5]
```diff
@@ -3512,3 +3527,390 @@ def generate_slurm_scripts(slurm_config_filename, popsycle_config_filename,
...
+        The maximum radius from the galactic center that you want to find PBHs at.
+        Defaults to 8.3 kpc (Where Earth is located)
+
+    c : float
+        Concentration index.
+        Defaults to 10 (concentration index of the Milky Way, multiple papers show it to be anywhere from 10-12)
+
+    r_vir : float
+        The virial radius.
+        Defaults to 200 kpc (The virial radius of the Milky Way)
```
Comment (MichaelMedford) [inline popsycle/synthetic.py]: If these are stated as having default values, then they need to have defaults assigned to them in the arguments of `add_pbh`

File: popsycle/synthetic.py [hunk 6]
```diff
@@ -3512,3 +3527,390 @@ def generate_slurm_scripts(slurm_config_filename, popsycle_config_filename,
...
+
+    #Define parameters for NFW profile calculations
+    r_s = r_vir/c #kpc, scale radius
+    g = 4.3*(10**-3) #(pc*km^2)/(Msun*s^2)
+    h = 70 #km/(s*Mpc)
+
+    #Read in the hdf5 file that doesn't have PBHs. Product of perform_pop_syn.
+    no_pbh_hdf5_file = h5py.File(hdf5_file)
+    key_list = list(no_pbh_hdf5_file)
+    key_list = key_list[:len(key_list)-2]
```
Comment (MichaelMedford) [inline popsycle/synthetic.py]: Is this meant to remove the 'lat_bin_edges' and 'lon_bin_edges' keys? If so, then I would rather see something more explicit, like:
`key_list = [key for key in key_list if 'bin_edges' not in key]`

File: popsycle/synthetic.py [hunk 7]
```diff
@@ -3512,3 +3527,390 @@ def generate_slurm_scripts(slurm_config_filename, popsycle_config_filename,
...
+    field_of_view_area = 2*((area/np.pi)**(1/2))
+
+    #NFW Profile calculations to determine mass of dark matter within given distance of galactic center
+    rho_crit = ((3*(h**2))/(8*np.pi*g))*(10**6) #Msun/Mpc^3
+    rho_knot = (200/3)*(rho_crit)*((c**3)/(np.log(c+1)-(c/(1+c)))) # Msun/Mpc^3
+    mass_8_kpc = (4*np.pi*rho_knot*(r_s**3)*(np.log((r_s + r_max)/r_s) - (r_max/(r_s+r_max))))*(((10**3)**3)/((10**6)**3))*fdm #Msun
+
+    #Determine the number of PBHs within that distance
+    num_pbh_8kpc = (mass_8_kpc/pbh_mass)
+    num_pbh_8kpc = round(num_pbh_8kpc, -1)
```
Comment (MichaelMedford) [inline popsycle/synthetic.py]: Why is there a -1 in this round?

File: popsycle/synthetic.py [hunk 8]
```diff
@@ -3512,3 +3527,390 @@ def generate_slurm_scripts(slurm_config_filename, popsycle_config_filename,
...
+
+    #Defining needed functions from the python package "NFWdist".
```
Comment (MichaelMedford) [inline popsycle/synthetic.py]: Please label these with more comments. I do not understand what these functions are doing or how they work.

File: popsycle/synthetic.py [hunk 9]
```diff
@@ -3512,3 +3527,390 @@ def generate_slurm_scripts(slurm_config_filename, popsycle_config_filename,
...
+        else:
+            if (p > 1):
+                p = 1
+            elif (p <= 0):
+                p = 0
+        if hasattr(con, '__len__'):
+            p = np.outer(p,pnfwunorm(1, con=con))
+        else:
+            p *= pnfwunorm(1, con=con)
+        return (-(1.0/np.real(special.lambertw(-np.exp(-p-1))))-1)/con
```
Comment (MichaelMedford) [inline popsycle/synthetic.py]: Please label these with more comments. I do not understand what these functions are doing or how they work.

File: popsycle/synthetic.py [hunk 10]
```diff
@@ -3512,3 +3527,390 @@ def generate_slurm_scripts(slurm_config_filename, popsycle_config_filename,
...
+    lat_bin = pd.DataFrame(np.array(no_pbh_hdf5_file['lat_bin_edges']))
+    long_bin = pd.DataFrame(np.array(no_pbh_hdf5_file['long_bin_edges']))
+    no_pbh_hdf5_file.close()
+
+    field_of_view_area = 2*((area/np.pi)**(1/2))
+
+    #NFW Profile calculations to determine mass of dark matter within given distance of galactic center
+    rho_crit = ((3*(h**2))/(8*np.pi*g))*(10**6) #Msun/Mpc^3
+    rho_knot = (200/3)*(rho_crit)*((c**3)/(np.log(c+1)-(c/(1+c)))) # Msun/Mpc^3
+    mass_8_kpc = (4*np.pi*rho_knot*(r_s**3)*(np.log((r_s + r_max)/r_s) - (r_max/(r_s+r_max))))*(((10**3)**3)/((10**6)**3))*fdm #Msun
```
Comment (MichaelMedford) [inline popsycle/synthetic.py]: This variable, and all others with `8_kpc` in the name, should be renamed, as the radius within which the PBHs are calculated is a variable.

File: popsycle/synthetic.py [hunk 11]
```diff
@@ -3512,3 +3527,390 @@ def generate_slurm_scripts(slurm_config_filename, popsycle_config_filename,
...
+    #Determining the center of the field of view circle.
+    l_min = np.min(long_bin).values
+    l_max = np.max(long_bin).values
+    b_min = np.min(lat_bin).values
+    b_max = np.max(lat_bin).values
+    l_mid = np.average((l_min, l_max))
+    b_mid = np.average((b_min, b_max))
+
+    dists = angdist(l_mid, b_mid, c.l.deg, c.b.deg)
+    r_max_mask = 16.6
```
Comment (MichaelMedford) [inline popsycle/synthetic.py]: What is this here for?

File: popsycle/synthetic.py [hunk 12]
```diff
@@ -3512,3 +3527,390 @@ def generate_slurm_scripts(slurm_config_filename, popsycle_config_filename,
...
+    z=v_esc/v_knot
+    n_esc=scipy.special.erf(z)-(2*(np.pi**(-1/2))*z*(np.exp(-z**2)))
+    v=np.arange(0,525, 525/len(n_esc))
+
+    #Getting velocity for each PBH using formulas from Ranjan Laha 2018.
+    for disp in pbh_dispersion:
+        f = ((3/(2*np.pi*(disp**2)))**(3/2))*(np.exp((-(v**2))/(2*(disp**2))))
+        cdf = (3*(3**(1/2))*(1/(disp**2))*disp*scipy.special.erf(v/((2**(1/2))*disp)))/(4*np.pi)
+        rand_cdf = np.random.uniform(0, np.amax(cdf), 1)
+        interpreted_rms_velocities = np.interp(rand_cdf, cdf, v)
```
Comment (MichaelMedford) [inline popsycle/synthetic.py]: This loop assigned the value of `np.interp(rand_cdf, cdf, v)` to `interpreted_rms_velocities` for each value of `disp`. However what you are left with at the end of the loop is a single value for `interpreted_rms_velocities` equal to the last value of `pbh` in the loop. What you want to do instead is build up an array of values that get passed into `spherical_to_cartesian` below. 

Something like:
```
#Getting velocity for each PBH using formulas from Ranjan Laha 2018.

for disp in pbh_dispersion:
    f = ((3/(2*np.pi*(disp**2)))**(3/2))*(np.exp((-(v**2))/(2*(disp**2))))
    cdf = (3*(3**(1/2))*(1/(disp**2))*disp*scipy.special.erf(v/((2**(1/2))*disp)))/(4*np.pi)
    rand_cdf = np.random.uniform(0, np.amax(cdf), 1)
    interpreted_rms_velocities.appedn(np.interp(rand_cdf, cdf, v))
```

File: popsycle/synthetic.py [hunk 13]
```diff
@@ -3512,3 +3527,390 @@ def generate_slurm_scripts(slurm_config_filename, popsycle_config_filename,
...
+    #Defining some parameters that are used in the velocity distribution equations
+    v_esc = 525 #km/s
+    v_knot = ((2/3)**(1/2))*pbh_dispersion
+    z=v_esc/v_knot
+    n_esc=scipy.special.erf(z)-(2*(np.pi**(-1/2))*z*(np.exp(-z**2)))
+    v=np.arange(0,525, 525/len(n_esc))
+
+    #Getting velocity for each PBH using formulas from Ranjan Laha 2018.
+    for disp in pbh_dispersion:
+        f = ((3/(2*np.pi*(disp**2)))**(3/2))*(np.exp((-(v**2))/(2*(disp**2))))
```
Comment (MichaelMedford) [inline popsycle/synthetic.py]: What is `f` doing here? I don't see it used anywhere.

File: popsycle/synthetic.py [hunk 14]
```diff
@@ -3512,3 +3527,390 @@ def generate_slurm_scripts(slurm_config_filename, popsycle_config_filename,
...
+                'or pick a new name.')
+
+    # Error handling/complaining if input types are not right.
+
+    if type(output_root2) != str:
+        raise Exception('output_root must be a string.')
+
+    if type(galaxia_area) != float:
+        if type(galaxia_area) != int:
+            raise Exception('galaxia_area must be a float or an integer.')
```
Comment (MichaelMedford) [inline popsycle/synthetic.py]: Parameter is called area, not galaxia_area

File: popsycle/synthetic.py [hunk 15]
```diff
@@ -3512,3 +3527,390 @@ def generate_slurm_scripts(slurm_config_filename, popsycle_config_filename,
...
+
+    t0 = time.time()
+
+    #Define parameters for NFW profile calculations
+    r_s = r_vir/c #kpc, scale radius
+    g = 4.3*(10**-3) #(pc*km^2)/(Msun*s^2)
+    h = 70 #km/(s*Mpc)
+
+    #Read in the hdf5 file that doesn't have PBHs. Product of perform_pop_syn.
+    no_pbh_hdf5_file = h5py.File(hdf5_file)
```
Comment (MichaelMedford) [inline popsycle/synthetic.py]: Open with 'r' flag to avoid warning

File: popsycle/synthetic.py [hunk 16]
```diff
@@ -3512,3 +3527,390 @@ def generate_slurm_scripts(slurm_config_filename, popsycle_config_filename,
...
+    key_list = list(no_pbh_hdf5_file)
+    key_list = key_list[:len(key_list)-2]
+    lat_bin = pd.DataFrame(np.array(no_pbh_hdf5_file['lat_bin_edges']))
+    long_bin = pd.DataFrame(np.array(no_pbh_hdf5_file['long_bin_edges']))
+    no_pbh_hdf5_file.close()
+
+    field_of_view_area = 2*((area/np.pi)**(1/2))
```
Comment (MichaelMedford) [inline popsycle/synthetic.py]: This is a diameter and should be renamed to not have the word 'area' in the variable

File: popsycle/synthetic.py [hunk 17]
```diff
@@ -3512,3 +3527,390 @@ def generate_slurm_scripts(slurm_config_filename, popsycle_config_filename,
...
+    Output:
+    distance [degrees] angular distance between the two points
+    '''
+    from numpy import pi, sin, cos, arccos
+        d2r = pi / 180.0
+        cosdist = (cos(np.absolute(ra1 - ra2) * d2r) * cos(np.absolute(dec1 * d2r)) * cos(np.absolute(dec2 * d2r)) +
+                   sin(np.absolute(dec1 * d2r)) * sin(np.absolute(dec2 * d2r)))
+        distance = arccos(cosdist) / d2r
+        return distance
+
```
Comment (MichaelMedford) [inline popsycle/synthetic.py]: This has an indentation error and will not run as written

File: popsycle/synthetic.py [hunk 18]
```diff
@@ -3512,3 +3527,390 @@ def generate_slurm_scripts(slurm_config_filename, popsycle_config_filename,
...
+    b_min = np.min(lat_bin).values
+    b_max = np.max(lat_bin).values
+    l_mid = np.average((l_min, l_max))
+    b_mid = np.average((b_min, b_max))
+
+    dists = angdist(l_mid, b_mid, c.l.deg, c.b.deg)
+    r_max_mask = 16.6
+
+    #Masking the full PBH data for our field of view
+    mask = (c.l.deg >= l_min) & (c.l.deg <= l_max) & (c.b.deg >= b_min) & (c.b.deg <= b_max) & (c.distance.kpc <= r_max_mask) & (dists < field_of_view_area)
```
Comment (MichaelMedford) [inline popsycle/synthetic.py]: I do not understand this line. We are masking for PBHs that fall within the cone of our observation. Why do we ask for the box of (l_min to l_max) by (b_min to b_max) if we are also asking for (dists < field_of_view_area) where dists are the separation from the center of the field? Why is field_of_view_area a diameter instead of a radius? And why do we query for distances twice of r_max instead of inside of r_max?

File: popsycle/synthetic.py (hunk 4 above)
Comment (Kerianne28) [reply popsycle/synthetic.py]: I kept getting an error that said "array is not C contiguous". After nearly a full day of trying everything I could this was the "easiest" fix. No one else seemed to be getting this issue. I couldn't figure it out for the life of me.

File: popsycle/synthetic.py [hunk 19]
```diff
@@ -3512,3 +3527,390 @@ def generate_slurm_scripts(slurm_config_filename, popsycle_config_filename,
...
+    long_vel = np.random.uniform(0, 2*np.pi, len(data_in_field))
+
+    #Transforming velocities to cartesian to get vx, vy, vz values
+    cart_vel = astropy.coordinates.spherical_to_cartesian(interpreted_rms_velocities, lat_vel, long_vel)
+    vx= cart_vel[0]
+    vy= cart_vel[1]
+    vz= cart_vel[2]
+
+    #Getting rest of data for PBHs for the combined .h5 file
+    mass = np.full((len(data_in_field), 1), pbh_mass)
```
Comment (MichaelMedford) [inline popsycle/synthetic.py]: This line, and all subsequent examples of calling `np.full` followed by `np.reshape`, can be changed to `mass = np.full(len(data_in_field), pbh_mass)`

File: popsycle/synthetic.py (hunk 5 above)
Comment (Kerianne28) [reply popsycle/synthetic.py]: These defaults are listed in the pbh_config.yaml file. I was under the impression that if you didn't specify one of the values, it would pull it from the config file. Is the not the case? Do I still need to assign them in add_pbh? And if I do still need to assign them in add_pbh, what is the point of pbh_config.yaml? (As I thought that was its function, I am confused now).

File: popsycle/synthetic.py [hunk 20]
```diff
@@ -3512,3 +3527,390 @@ def generate_slurm_scripts(slurm_config_filename, popsycle_config_filename,
...
+    ubv_b = np.reshape(ubv_b, (len(data_in_field),))
+    ubv_h = np.full((len(data_in_field), 1), np.nan)
+    ubv_h = np.reshape(ubv_h, (len(data_in_field),))
+    ubv_v = np.full((len(data_in_field), 1), np.nan)
+    ubv_v = np.reshape(ubv_v, (len(data_in_field),))
+
+    pbh_data = pd.DataFrame({'zams_mass':zams_mass, 'rem_id':rem_id, 'mass':mass, 'px':px, 'py':py, 'pz':pz, 'vx':vx, 'vy':vy, 'vz':vz, 'rad':r, 'glat':b, 'glon':l, 'vr':vr, 'mu_b':mu_b, 'mu_lcosb':mu_lcosb, 'age':age, 'popid':pop_id, 'ubv_k':ubv_k, 'ubv_i':ubv_i, 'exbv':exbv, 'obj_id':obj_id, 'ubv_j':ubv_j, 'ubv_u':ubv_u, 'ubv_r':ubv_r, 'ubv_b':ubv_b, 'ubv_h':ubv_h, 'ubv_v':ubv_v})
+
+    #Getting information out of .h5 file before PBHs
+    no_pbh_hdf5_file = h5py.File(hdf5_file)
```
Comment (MichaelMedford) [inline popsycle/synthetic.py]: Add 'r' flag to this open to avoid a warning

File: popsycle/synthetic.py (hunk 7 above)
Comment (Kerianne28) [reply popsycle/synthetic.py]: -1 rounds it to the nearest 10. There were some issues that came up when not rounded to the nearest 10. That was when the code was slightly different. I'll check if it is still needed.

File: popsycle/synthetic.py (hunk 11 above)
Comment (Kerianne28) [reply popsycle/synthetic.py]: I will change it to 2*r_max. Thank you. It is twice r_max because at the point of the mask the radii are in galactic/heliocentric coordinates. Will said that we need to consider objects further than the galactic center as there could be sources or lenses back there.

File: popsycle/synthetic.py [hunk 21]
```diff
@@ -3512,3 +3527,390 @@ def generate_slurm_scripts(slurm_config_filename, popsycle_config_filename,
...
+
+    #Converting spherical galactocentric coordinates to cartesian galactocentric coordinates
+    cart = astropy.coordinates.spherical_to_cartesian(r_values, lats, longs)
+
+    #Defining galactocentric coordinates
+    c = coord.Galactocentric(x=cart[0] * units.kpc, y=cart[1] * units.kpc, z=cart[2] * units.kpc)
+
+    #Transforming from galactocentric to galactic coordinates
+    #outputs l, b, dist in degrees.
+    c = c.transform_to(coord.Galactic(representation_type='cartesian'))
```
Comment (MichaelMedford) [inline popsycle/synthetic.py]: Check if this comes out with longitudes from 0-360 or from -180 to 180. We use the convention of -180 to 180 in the hdf5 file, and I believe that astropy uses the 0-360 convention.

File: popsycle/synthetic.py [hunk 22]
```diff
@@ -3512,3 +3527,390 @@ def generate_slurm_scripts(slurm_config_filename, popsycle_config_filename,
...
+
+    #Set minimum and maximum l, b, and r values for PBH mask. 
+    #Determining the center of the field of view circle.
+    l_min = np.min(long_bin).values
+    l_max = np.max(long_bin).values
+    b_min = np.min(lat_bin).values
+    b_max = np.max(lat_bin).values
+    l_mid = np.average((l_min, l_max))
+    b_mid = np.average((b_min, b_max))
```
Comment (MichaelMedford) [inline popsycle/synthetic.py]: `l_mid` and `b_mid` should be passed in as parameters, given that they are also in the galaxia file and the field_config file. Depending on how the lat_bin and long_bins are created, the average may not always equal the center of the field.

File: popsycle/synthetic.py (hunk 12 above)
Comment (Kerianne28) [reply popsycle/synthetic.py]: Even though I updated the velocities I am having this same issue. Working on it now.

File: popsycle/synthetic.py [hunk 23]
```diff
@@ -3512,3 +3527,394 @@ def generate_slurm_scripts(slurm_config_filename, popsycle_config_filename,
...
+
+    pbh_vrms = np.interp(pbh_r_galacto, vel_data['x'], vel_data['y'])
+    v_vals = np.arange(0, v_esc) #Goes from v to v_esc
+    a = (1/2)*pbh_vrms*((np.pi/2)**(1/2))
+
+    for a_val in a:
+        pdf = ((2/np.pi)**(1/2))*((v_vals**2*np.exp(-v_vals**2/(2*a_val**2)))/a_val**3)
+        cdf = scipy.special.erf(v_vals/(a_val*2**(1/2)))-(((2/np.pi)**(1/2))*((v_vals*np.exp(-v**2/2*a_val**2))/a_val))
+        rand_cdf = np.random.uniform(0, np.amax(cdf), 1)
+        interpreted_rms_velocities = np.interp(rand_cdf, cdf, v_vals)
```
Comment (MichaelMedford) [inline popsycle/synthetic.py]: interpreted_rms_velocities should be an array that is filled up on each iteration of the loop

File: popsycle/synthetic.py (hunk 12 above)
Comment (Kerianne28) [reply popsycle/synthetic.py]: Fixed in next version

File: popsycle/synthetic.py (hunk 23 above)
Comment (Kerianne28) [reply popsycle/synthetic.py]: Fixed. It will be reflected in next push.

**Comment by jluastro** [comment]
Actually, the way it is currently implemented is slightly faster. But I would recommend documenting that step a little more. 

I also just thought of this: do all hdf5 implementations preserve key order? If not, then Michael’s suggested implementation is preferred or perhaps instead just do:

key_list.remove(‘lat_bin_edges’)
key_list.remove(‘lon_bin_edges’)

This will throw an error if they don’t exist; but you can catch it if needed.

**Comment by MichaelMedford** [comment]
I would be surprised if the key order weren't preserved given that the h5py
is wrapped around hdf5, that is allocating memory in an order. But I
couldn't find documentation on it. My suggestion was motivated by wanting
to (1) guarantee correct behavior and (2) improve readability.

File: popsycle/synthetic.py (hunk 21 above)
Comment (Kerianne28) [reply popsycle/synthetic.py]: Can I literally just subtract 180 from all of the longitudes to fix this?

File: popsycle/synthetic.py (hunk 18 above)
Comment (Kerianne28) [reply popsycle/synthetic.py]: I went ahead and made the mask just on (dists < field_of_view_area). Again, as far as the 2*r_max is concerned, Will said we need to consider objects further away than the bulge. And it did make a difference.

File: popsycle/synthetic.py [hunk 24]
```diff
@@ -3512,3 +3527,402 @@ def generate_slurm_scripts(slurm_config_filename, popsycle_config_filename,
...
+        vars()[key+'_max_l'] = np.amax(keys.T[11])
+        vars()[key+'_min_l'] = np.amin(keys.T[11])
+        vars()[key+'_max_b'] = np.amax(keys.T[10])
+        vars()[key+'_min_b'] = np.amin(keys.T[10])
+        vars()[key+'_mask'] = (pbh_data['glon'] >= vars()[key+'_min_l']) & (pbh_data['glon'] <= vars()[key+'_max_l']) & (pbh_data['glat'] >= vars()[key+'_min_b']) & (pbh_data['glat'] <= vars()[key+'_max_b'])
+        vars()['pbh_'+key] = (pbh_data[vars()[key+'_mask']]).T
+        vars()['pbh_'+key].reset_index(drop=True, inplace=True)
+        vars()['full_'+key]=pd.concat([keys, vars()['pbh_'+key]], axis=1)
+    no_pbh_hdf5_file.close()
+
```
Comment (MichaelMedford) [inline popsycle/synthetic.py]: What we want to do is take each key in the hdf5 file, and append rows for all of the PBHs that this function generates. What this section of code does is add new keys with PBH information. Look at the documentation for perform_pop_syn and make sure that you have the right information in the right columns and replace the keys.

File: popsycle/synthetic.py [hunk 25]
```diff
@@ -3512,3 +3527,402 @@ def generate_slurm_scripts(slurm_config_filename, popsycle_config_filename,
...
+    px = cart_helio[0]
+    py = cart_helio[1]
+    pz = cart_helio[2]
+
+    vr, mu_b, mu_lcosb = calc_sph_motion(vx, vy, vz, r_in_field, b_in_field, l_in_field)
+
+    exbv = np.full(len(data_in_field), 0)
+    ubv_k = np.full(len(data_in_field), np.nan)
+    ubv_j = np.full(len(data_in_field), np.nan)
+    obj_id = np.arange(1, (len(data_in_field)+1))
```
Comment (MichaelMedford) [inline popsycle/synthetic.py]: Will this object id collide with others in the dataset? I think that it will. Come up with a convention that makes sure that all object ids are unique.

File: popsycle/synthetic.py [hunk 26]
```diff
@@ -3512,3 +3527,402 @@ def generate_slurm_scripts(slurm_config_filename, popsycle_config_filename,
...
+
+    cart_helio = astropy.coordinates.spherical_to_cartesian(r_in_field, b_rad, l_rad)
+
+    px = cart_helio[0]
+    py = cart_helio[1]
+    pz = cart_helio[2]
+
+    vr, mu_b, mu_lcosb = calc_sph_motion(vx, vy, vz, r_in_field, b_in_field, l_in_field)
+
+    exbv = np.full(len(data_in_field), 0)
```
Comment (MichaelMedford) [inline popsycle/synthetic.py]: Why 0?

File: popsycle/synthetic.py [hunk 27]
```diff
@@ -3512,3 +3527,402 @@ def generate_slurm_scripts(slurm_config_filename, popsycle_config_filename,
...
+    #Transforming velocities to cartesian to get vx, vy, vz values
+    cart_vel = astropy.coordinates.spherical_to_cartesian(interpreted_rms_velocities, lat_vel, long_vel)
+    vx= cart_vel[0]
+    vy= cart_vel[1]
+    vz= cart_vel[2]
+
+    #Getting rest of data for PBHs for the combined .h5 file
+    mass = np.full(len(data_in_field), pbh_mass)
+    zams_mass = mass
+    age = np.full(len(data_in_field), 14*(10**9))
```
Comment (MichaelMedford) [inline popsycle/synthetic.py]: Why this age?

File: popsycle/synthetic.py [hunk 28]
```diff
@@ -3512,3 +3527,402 @@ def generate_slurm_scripts(slurm_config_filename, popsycle_config_filename,
...
+        vel_data = pd.read_csv('data/inner_slope_half.csv')
+
+    pbh_vrms = np.interp(pbh_r_galacto, vel_data['x'], vel_data['y'])
+    v_vals = np.arange(0, v_esc) #Goes from v to v_esc
+    a = (1/2)*pbh_vrms*((np.pi/2)**(1/2))
+
+    rand_cdf = []
+
+    for a_val in a:
+        pdf = ((2/np.pi)**(1/2))*((v_vals**2*np.exp(-v_vals**2/(2*a_val**2)))/a_val**3)
```
Comment (MichaelMedford) [inline popsycle/synthetic.py]: Is this line still needed? I don't see where it is needed.

File: popsycle/synthetic.py (hunk 4 above)
Comment (MichaelMedford) [reply popsycle/synthetic.py]: You cannot do this. It has to be changed back because we do not know how it effects the results of the query.

File: popsycle/synthetic.py [hunk 29]
```diff
@@ -3758,133 +3773,151 @@ def qnfw(p, con=5, logp=False):
...
-    d_lat = f.create_dataset('lat_bin_edges', (len(lat_bin), 1), data=lat_bin)
-    d_long = f.create_dataset('long_bin_edges', (len(lat_bin), 1), data=long_bin)
-    f.close()
+    t1 = time.time()
+    print('Total runtime: {0:f} s'.format(t1 - t0))
 
     return
 
-def generate_pbh_config_file(config_filename, fdm, pbh_mass, r_max, c, r_vir):
+def generate_pbh_config_file(config_filename, fdm, pbh_mass, r_max, c, r_vir, inner_slope, v_esc):
```
Comment (MichaelMedford) [inline popsycle/synthetic.py]: Either set default values in this definition, or remove the language about default values in the docstring below.

File: popsycle/run.py [hunk 30]
```diff
@@ -95,7 +95,14 @@ def run():
...
                           help="Overwrite all output files.",
                           action='store_true')
-    args = parser.parse_args()
+
+
+    optional.add_argument('--pbh-config-filename', type=str,
+                      help='Name of configuration file containing '
+                           'pbh inputs. Default if needed is: '
+                           'pbh_config.yaml',
+                      default='pbh_config.yaml')
```
Comment (MichaelMedford) [inline popsycle/run.py]: Is this argument has a default value, then the condition on Line 186 (if args.pbh_config_filename is not None) will always be True.  Let's remove the default here.

File: popsycle/synthetic.py [hunk 31]
```diff
@@ -3512,3 +3528,438 @@ def generate_slurm_scripts(slurm_config_filename, popsycle_config_filename,
...
+    l_min = np.min(long_bin).values
+    l_max = np.max(long_bin).values
+    b_min = np.min(lat_bin).values
+    b_max = np.max(lat_bin).values
+
+    #Calculating distances between each latitude, longitude pair, and the center of the field of view.
+    dists = angdist(l, b, longitude, latitude)
+
+    #Masking the full PBH data to obtain just the PBHs in our field of view.
+    mask = (galactic.distance.kpc <= 2*r_max) & (dists < field_of_view_diameter)
```
Comment (MichaelMedford) [inline popsycle/synthetic.py]: why is this mask for dists less than diameter instead of radius?

File: popsycle/synthetic.py [hunk 32]
```diff
@@ -3512,3 +3528,438 @@ def generate_slurm_scripts(slurm_config_filename, popsycle_config_filename,
...
+    longitude=galactic.l.deg
+    #Adjusting longitude values to match the coordinate format that we need
+    longitude=np.where(longitude>180, longitude-360, longitude)   
+    
+    #Set minimum and maximum l and b for PBH mask. 
+    l_min = np.min(long_bin).values
+    l_max = np.max(long_bin).values
+    b_min = np.min(lat_bin).values
+    b_max = np.max(lat_bin).values
+
```
Comment (MichaelMedford) [inline popsycle/synthetic.py]: are these still needed?

File: popsycle/synthetic.py [hunk 33]
```diff
@@ -3512,3 +3528,438 @@ def generate_slurm_scripts(slurm_config_filename, popsycle_config_filename,
...
+    cart_pbh = astropy.coordinates.cartesian_to_spherical(galactic_pbh.x, galactic_pbh.y, galactic_pbh.z)
+    pbh_r_galacto = cart_pbh[0]
+    
+    #Inner slope of the MW halo
+    #From Lacroix et al 2018, Figure 11 (top left panel)
+    if inner_slope == 1:
+        vel_data = pd.read_csv('data/radial_velocity_profile_steep.csv')
+    elif inner_slope == .25:
+        vel_data = pd.read_csv('data/radial_velocity_profile_shallow.csv')
+    else:
```
Comment (MichaelMedford) [inline popsycle/synthetic.py]: What if inner_slope = 5? Or None? Or a string? It's not safe to have a catch-all "else" here. Replace this with another "elif" and then raise an exception and exit upon the else.
